transport/shadowtls/shadowtls.go

```diff
@@ -0,0 +1,146 @@
+package shadowtls
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"io"
+	"net"
+
+	"github.com/Dreamacro/clash/common/pool"
+	"github.com/Dreamacro/clash/transport/trojan"
+)
+
+const (
+	chunkSize           = 1 << 13
+	Mode         string = "shadow-tls"
+	hashLen      int    = 8
+	tlsHeaderLen int    = 5
+)
+
+// TLSObfs is shadowsocks tls simple-obfs implementation
+type ShadowTls struct {
+	net.Conn
+	password     []byte
+	remain       int
+	firstRequest bool
+	tlsConnector *trojan.Trojan
+}
+
+type HashedConn struct {
+	net.Conn
+	hasher hash.Hash
+}
+
+func newHashedStream(conn net.Conn, password []byte) HashedConn {
+	return HashedConn{
+		Conn:   conn,
+		hasher: hmac.New(sha1.New, password),
+	}
+}
+
+func (h HashedConn) Read(b []byte) (n int, err error) {
+	n, err = h.Conn.Read(b)
+	h.hasher.Write(b[:n])
+	return
+}
+
+func (to *ShadowTls) read(b []byte) (int, error) {
+	buf := pool.Get(tlsHeaderLen)
+	_, err := io.ReadFull(to.Conn, buf)
+	if err != nil {
+		return 0, fmt.Errorf("shadowtls read failed %w", err)
+	}
+	if buf[0] != 0x17 || buf[1] != 0x3 || buf[2] != 0x3 {
+		return 0, fmt.Errorf("invalid shadowtls header %v", buf)
+	}
+	length := int(binary.BigEndian.Uint16(buf[3:]))
+	pool.Put(buf)
+
+	if length > len(b) {
+		n, err := to.Conn.Read(b)
+		if err != nil {
+			return n, err
+		}
+		to.remain = length - n
+		return n, nil
+	}
+
+	return io.ReadFull(to.Conn, b[:length])
+}
+
+func (to *ShadowTls) Read(b []byte) (int, error) {
+	if to.remain > 0 {
+		length := to.remain
+		if length > len(b) {
+			length = len(b)
+		}
+
+		n, err := io.ReadFull(to.Conn, b[:length])
+		if err != nil {
+			return n, fmt.Errorf("shadowtls Read failed with %w", err)
+		}
+		to.remain -= n
+		return n, nil
+	}
+
+	return to.read(b)
+}
+
+func (to *ShadowTls) Write(b []byte) (int, error) {
+	length := len(b)
+	for i := 0; i < length; i += chunkSize {
+		end := i + chunkSize
+		if end > length {
+			end = length
+		}
+
+		n, err := to.write(b[i:end])
+		if err != nil {
+			return n, fmt.Errorf("shadowtls Write failed with %w, i=%d, end=%d, n=%d", err, i, end, n)
+		}
+	}
+	return length, nil
+}
+
+func (s *ShadowTls) write(b []byte) (int, error) {
+	var hashVal []byte
+	if s.firstRequest {
+		hashedConn := newHashedStream(s.Conn, s.password)
+		if _, err := s.tlsConnector.StreamConn(hashedConn); err != nil {
+			return 0, fmt.Errorf("tls connect failed with %w", err)
+		}
+		hashVal = hashedConn.hasher.Sum(nil)[:hashLen]
+		s.firstRequest = false
+	}
+
+	buf := pool.GetBuffer()
+	defer pool.PutBuffer(buf)
+	buf.Write([]byte{0x17, 0x03, 0x03})
+	binary.Write(buf, binary.BigEndian, uint16(len(b)+len(hashVal)))
+	buf.Write(hashVal)
+	buf.Write(b)
+	remain := buf.Len()
+	for remain > 0 {
+		n, err := s.Conn.Write(buf.Bytes())
```

Comment:
Do you mean `n, err := s.Conn.Write(buf.Bytes()[buf.Len()-remain:])`?  
btw, I prefer this:  
```go
// start from L123
buf.Write(b)
n, err := s.Conn.Write(buf.Bytes())
if frontHeadroom := tlsHeaderLen + len(hashVal); n >= frontHeadroom {
	n -= frontHeadroom
} else {
	n = 0
}
return n, error
```

Reply:
> Do you mean n, err := s.Conn.Write(buf.Bytes()[buf.Len()-remain:])?

outdated XD

Reply:
Thanks for pointing it out. I've made some change since this version.
[The docs](https://pkg.go.dev/io#Writer) said a writer must return an error if n < len(buf).
I kinda suspect if it's worth it to return the exact number of bytes written since we'll return an error anyways.
Another reason I would prefer just return 0 is that once we have a partially written packet, the connection is poisoned.
For example, the packet is 100 bytes long and we only write 50 bytes. The next call to `write` could try to write the remaining 50 bytes, but `write` will treat it as next packet (and add header (0x17...) to it). The server would take this as the remaining 50 bytes (it doesn't know that actually comes with a header & it expects 100 bytes) then the connection has doomed to be closed.
So in this case simply return a `0, err` would probably enough.

Reply:
> im outdated sry
LOL no worries at all!

Reply:
Yes, you are right.